Mathlib/Tactic/Peel.lean

```diff
@@ -0,0 +1,215 @@
+      goal.assign (← mkAppM ``Filter.Frequently.mp #[e, event_forall])
+      let (fvars, new_goal) ← all_imp.mvarId!.introN 2 [n.getD n₂, n'.getD freshName]
+      return (fvars[1]!, [new_goal])
+    | _, _ =>
+      match ty, target with
+        | .forallE _ t₁ b₁ _, .forallE n₂ t₂ b₂ c => do
+          unless ← isDefEq (← whnfR t₁) (← whnfR t₂) do
+            return (.none, [goal])
+          let all_imp ← mkFreshExprMVar <| ← withoutModifyingState <|
+            withLocalDecl n₂ c t₂ fun x => do
+              mkForallFVars #[x] (← mkArrow (b₁.instantiate1 x) (b₂.instantiate1 x))
+          goal.assign (← mkAppM ``forall_imp #[all_imp, e])
+          let (fvars, new_goal) ← all_imp.mvarId!.introN 2 [n.getD n₂, n'.getD freshName]
+          return (fvars[1]!, [new_goal])
+        | _, _ => return (.none, [goal])
+
+/-- Peels `n` quantifiers off the expression `e` and the main goal without naming the introduced
+variables. The expression `e`, with quantifiers removed, is assigned the default name `this`. -/
+def peelNum (e : Expr) (n : Nat) : TacticM Unit := withMainContext do
+  match n with
+    | 0 => pure ()
+    | 1 => let _ ← liftMetaTacticAux (peelQuantifier · e (n' := `this))
+    | n + 2 =>
+      let fvar? ← liftMetaTacticAux (peelQuantifier · e (n' := `this))
+      if let some fvarId := fvar? then
+        peelNum (.fvar fvarId) (n + 1)
+        let mvarId ← (← getMainGoal).clear fvarId
+        replaceMainGoal [mvarId]
+
+/-- Given a list `l` of names, this continues to peel quantifiers off of the expression `e` and
+the main goal and introduces variables with the provided names until the list of names is exhausted.
+Note: the first name in the list is used for the name of the expression `e` with quantifiers
+removed. If `l` is empty, one quantifier is removed with the default name `this`. -/
+def peelArgs (e : Expr) (l : List Name) : TacticM Unit := withMainContext do
+  match l with
+    | [] => let _ ← liftMetaTacticAux (peelQuantifier · e (n' := `this))
+    | [h] => let _ ← liftMetaTacticAux (peelQuantifier · e (n' := h))
+    | [h₁, h₂] => let _ ← liftMetaTacticAux (peelQuantifier · e h₂ h₁)
+    | h₁ :: h₂ :: h₃ :: hs =>
+      let fvar? ← liftMetaTacticAux (peelQuantifier · e h₂ h₁)
+      if let some fvarId := fvar? then
+        peelArgs (.fvar fvarId) (h₁ :: h₃ :: hs)
+        let mvarId ← (← getMainGoal).clear fvarId
+        replaceMainGoal [mvarId]
+
+private lemma Filter.frequently_congr {α : Type*} {f : Filter α} {p : α → Prop} {q : α → Prop}
+    (h : ∀ᶠ (x : α) in f, p x ↔ q x) : (∃ᶠ (x : α) in f, p x) ↔ ∃ᶠ (x : α) in f, q x := by
+  constructor
+  all_goals refine (Frequently.mp · (h.mono fun _ => ?_))
+  exacts [Iff.mp, Iff.mpr]
+
+/-- Peel off a single quantifier from an `↔`. -/
+def peelIffAux : TacticM Unit := withMainContext do
+  evalTactic (← `(tactic|
+    apply_rules [forall_congr', exists_congr, Filter.eventually_congr,
```

Comment:
`apply_rules` iteratively applies the lemmas, right? If you want to do a single iteration, you might have to write `first | apply forall_congr' | apply exists_congr | apply ...`

It would be good to have a useful error message if none apply. With the `first` tactic, you can have the last tactic be `fail "error message"`.